File: euler21.py
# Подсчитайте сумму всех дружественных чисел меньше 10000

from datetime import datetime
from math import ceil, sqrt


def all_factors_list(number, justcount=False):
    """Возвращает список делителей заданного числа, либо количество делителей"""
    divisor, divisor_list = 2, [1, number]
    for i in range(2, ceil(sqrt(number)) + 1):
        if number % i == 0:
            if justcount:
                divisor += 2
            else: 
                divisor_list.append(i)
                divisor_list.append(number // i)
    if justcount:
        return divisor - 1 if ceil(sqrt(number)) ** 2 == number else divisor
    else:
        return divisor_list[:-1] if ceil(sqrt(number)) ** 2 == number else divisor_list


def euler21():
    n = 10_000

    # 1 вариант
    start_time = datetime.now()
    amicable_pairs = []
    for x in range(10, n):
        if x not in amicable_pairs:
            x_sum_divisors = sum(all_factors_list(x)) - x
            y_sum_divisors = sum(all_factors_list(x_sum_divisors)) - x_sum_divisors
            if y_sum_divisors == x and x != x_sum_divisors:
                amicable_pairs.append(x)
                amicable_pairs.append(x_sum_divisors)
    print(sum(amicable_pairs))
    print(datetime.now() - start_time)

    # 2 вариант
    start_time = datetime.now()
    divisorsum = [0] * n
    for i in range(1, n):
        for j in range(i * 2, n, i):
            divisorsum[j] += i
    sum_amic_pairs = 0
    for i in range(1, n):
        j = divisorsum[i]
        if j != i and j < n and divisorsum[j] == i: sum_amic_pairs += i
    print(sum_amic_pairs)
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler21()

File: go/euler21.go
// euler21 считает сумму всех дружественных чисел меньше 10000.
//
// Решение задачи Эйлера №21.
package main

import (
	"fmt"
	"math"
	"time"
)

func allFactorsList(number int) []int {
	divisorList := []int{1, number}
	for i := 2; i <= int(math.Sqrt(float64(number)))+1; i++ {
		if number%i == 0 {
			divisorList = append(divisorList, i)
			divisorList = append(divisorList, number/i)
		}
	}
	return divisorList
}

func euler21() {
	n := 10000

	startTime := time.Now()
	amicablePairs := make([]int, 0)
	for x := 10; x < n; x++ {
		if !contains(amicablePairs, x) {
			xSumDivisors := sumSlice(allFactorsList(x)) - x
			ySumDivisors := sumSlice(allFactorsList(xSumDivisors)) - xSumDivisors
			if ySumDivisors == x && x != xSumDivisors {
				amicablePairs = append(amicablePairs, x)
				amicablePairs = append(amicablePairs, xSumDivisors)
			}
		}
	}
	fmt.Println(sumSlice(amicablePairs))
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	divisorSum := make([]int, n)
	for i := 1; i < n; i++ {
		for j := i * 2; j < n; j += i {
			divisorSum[j] += i
		}
	}
	sumAmicPairs := 0
	for i := 1; i < n; i++ {
		j := divisorSum[i]
		if j != i && j < n && divisorSum[j] == i {
			sumAmicPairs += i
		}
	}
	fmt.Println(sumAmicPairs)
	fmt.Println(time.Since(startTime))
}

func contains(slice []int, val int) bool {
	for _, v := range slice {
		if v == val {
			return true
		}
	}
	return false
}

func sumSlice(slice []int) int {
	total := 0
	for _, v := range slice {
		total += v
	}
	return total
}

func main() {
	euler21()
}
